server/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -205,40 +197,14 @@ public int hashCode() {
      * @param context Additional information needed to build the query
      */
     @SuppressWarnings("rawtypes")
-    public Query toFilter(ClusterService clusterService, ShardSearchRequest request, QueryShardContext context) {
+    public Query toFilter(ShardSearchRequest request, QueryShardContext context) {
         final MappedFieldType type = context.getFieldType(field);
         if (type == null) {
             throw new IllegalArgumentException("field " + field + " not found");
         }
 
-        int shardId = request.shardId().id();
-        int numShards = context.getIndexSettings().getNumberOfShards();
-        if (request.preference() != null || request.indexRoutings().length > 0) {
-            GroupShardsIterator<ShardIterator> group = buildShardIterator(clusterService, request);
-            assert group.size() <= numShards : "index routing shards: " + group.size() +
-                " cannot be greater than total number of shards: " + numShards;
-            if (group.size() < numShards) {
-                /*
-                 * The routing of this request targets a subset of the shards of this index so we need to we retrieve
-                 * the original {@link GroupShardsIterator} and compute the request shard id and number of
-                 * shards from it.
-                 */
-                numShards = group.size();
-                int ord = 0;
-                shardId = -1;
-                // remap the original shard id with its index (position) in the sorted shard iterator.
-                for (ShardIterator it : group) {
-                    assert it.shardId().getIndex().equals(request.shardId().getIndex());
-                    if (request.shardId().equals(it.shardId())) {
-                        shardId = ord;
-                        break;
-                    }
-                    ++ord;
-                }
-                assert shardId != -1 : "shard id: " + request.shardId().getId() + " not found in index shard routing";
-            }
-        }
-
+        int shardIndex = request.shardIndex() != -1 ? request.shardIndex() : request.shardId().id();
```

Comment:
I guess a case when `shardIdex() = -1` (when a coordinating node is v < 8.0)  and there is a `preference` or `indexRouting` present in scroll slice is a very rare case, and we can disregard it.

Reply:
Yes, that's my reasoning too.